pyra/hardware.py

```diff
@@ -0,0 +1,424 @@
+        dash_stats['network_sent'].append(network_sent_diff)
+
+    return network_received_diff, network_sent_diff
+
+
+def update():
+    """
+    Update all dashboard stats.
+
+    This function updates the cpu and memory usage of this python process as well as subprocesses. Following that the
+    system functions are called to update system cpu, gpu, memory, and network usage. Finally the keys in the
+    ``dash_stats`` dictionary are cleaned up to only hold 120 values. This function is called once per second, so
+    therefore there are 2 minutes worth of values in the dictionary.
+
+    Examples
+    --------
+    >>> update()
+    """
+    global initialized
+
+    current_timestamp = helpers.timestamp()
+
+    if initialized:
+        dash_stats['timestamp'].append(helpers.timestamp())
+
+        dash_stats['relative_time'] = []
+        for x in dash_stats['timestamp']:
+            seconds_ago = current_timestamp - x
+            dash_stats['relative_time'].append(seconds_ago)
+
+        child_processes = proc.children(recursive=False)  # list all children processes
+
+        for child in child_processes:
+            if child not in processes:
+                processes.append(child)
+
+    for p in processes:
+        # set the name
+        proc_name = definitions.Names().name if p.pid == proc_id else p.name()
+
+        # cpu stats per process
+        proc_cpu_percent = None
+        try:
+            proc_cpu_percent = min(100, p.cpu_percent())  # get current value, max of 100
+        except psutil.NoSuchProcess:
+            pass
+        finally:
+            if initialized:
+                try:
+                    dash_stats[f'cpu_{proc_name}']
+                except KeyError:
+                    dash_stats[f'cpu_{proc_name}'] = []
+                finally:  # append the current value to the list
+                    dash_stats[f'cpu_{proc_name}'].append(proc_cpu_percent)
+
+        # memory stats per process
+        proc_memory_percent = None
+        try:
+            proc_memory_percent = min(100, p.memory_percent(memtype='rss'))  # get current value, max of 100
+        except psutil.NoSuchProcess:
+            pass
+        finally:
+            if initialized:
+                try:
+                    dash_stats[f'memory_{proc_name}']
+                except KeyError:
+                    dash_stats[f'memory_{proc_name}'] = []
+                finally:  # append the current value to the list
+                    dash_stats[f'memory_{proc_name}'].append(proc_memory_percent)
+
+    update_cpu()  # todo, need to investigate why this is sometimes lower than the individual process
+    update_gpu()  # todo... AMD GPUs on non Linux... integrated GPUs... GPU stats for processes
+    update_memory()
+    update_network()  # todo... network stats for processes
+
+    for key in dash_stats:
+        dash_stats[key] = dash_stats[key][-history_length:]  # keep the first 2 minutes
```

Comment:
this could be a dict comprehension if you're into that kind of thing

Reply:
I changed this but still didn't use dict comprehension. I need to get more comfortable with that, might give it a try.

Reply:
They're awesome but somewhat unfriendly to less experienced python coders and easy to make incomprehensible to even experienced programmers. I see good arguments for and against their use in this case.